hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java

```diff
@@ -254,64 +293,105 @@ private void handlePlatformTransaction(
                     feeManager,
                     consensusNow);
 
-            // Now that we have a created handle context object and a consensus timestamp, run the appropriate {@code
-            // ConsensusTimeHook} event handlers
-            stakingPeriodTimeHook.process(consensusNow, context);
-            // @future('7836'): update the exchange rate and call from here
+            // Calculate the fee
+            fees = dispatcher.dispatchComputeFees(context);
 
-            // Dispatch the transaction to the handler
-            dispatcher.dispatchHandle(context);
+            // Run all pre-checks
+            final var preCheckResult = runPreChecks(consensusNow, verifier, preHandleResult);
+            if (preCheckResult.status() != SO_FAR_SO_GOOD) {
+                if (preHandleResult.status() == NODE_DUE_DILIGENCE_FAILURE) {
+                    payer = creator.accountId();
+                }
+                final var penaltyFee = new Fees(fees.nodeFee(), fees.networkFee(), 0L);
+                feeAccumulator.charge(payer, penaltyFee);
+                recordBuilder.status(preCheckResult.responseCodeEnum());
 
-            // TODO: Finalize transaction with the help of the token service
-            final var finalizationContext =
-                    new FinalizeContextImpl(preHandleResult.payer(), recordBuilder, configuration, stack);
-            transactionFinalizer.finalizeParentRecord(
-                    finalizationContext, List.of()); // TODO Need actual list of child records?
+            } else {
+                feeAccumulator.charge(payer, fees);
+                try {
+                    // Dispatch the transaction to the handler
+                    dispatcher.dispatchHandle(context);
+
+                    // commit state
+                    stack.commitFullStack();
+                    recordBuilder.status(SUCCESS);
+                } catch (final HandleException e) {
+                    rollback(e.getStatus(), stack, recordListBuilder);
+                    feeAccumulator.charge(payer, fees);
```

Comment:
I think there was an open question here as to whether in this case we also want to zero out the service fee. We should double check that.

Reply:
Let's leave it as is for now and see the result from testing